### PR TITLE
Support config.json and pass environment variables into ninja.build

### PR DESCRIPTION
**Pull request by rickychien**
According issue #31.

A scenario comes out of my mind to build Gaia would be:

```
DEBUG=1 MYPATH=path/to/somewhere bin/confidant . --config ./build-config.json
```

This patch is able to merge given variables such as DEBUG=1 MYPATH=path/to/somewhere and config variables inside build-config.json as a string of variables in order to pass them into build.ninja.

The idea of build-config.json can treat as a build config interface which should contain all build variables / flags and override by input / exports environment variables.

A build-config.json

``` json
{
  "DEBUG": 1,
  "MYPATH": "path/to/somewhere"
}
```

If user inputs variables that are inexistent in build-config.json that will not pass into build.ninja because they should be defined in build-config interface.

### CODE REVIEW

**Comment by lambdabaa** [comment]
Environment variables make sense to add support for. Special handling for config files does not. You can just have your `configure.js` depend on and read other configuration files.

**Comment by rickychien** [comment]
What's your mean for having your `confidant.js` depend on and read other configuration files?
The configuration files is optional for user if they want to have a quick setup and a good interface especially in the case of build system.

**Comment by rickychien** [comment]
Accodring to the spirit of confidant,

```
Configure your project for ninja's parallel and incremental builds with pure js.
```

I'm still thinking that communication between configure.js and ninja is supported to provide a solution to deliver and manage env variables. Therefore, I'd like to put nconf in confidant and hope that makes sense to you :)

**Comment by lambdabaa** [comment]
So what I meant was that there's no reason that we have to pass config files to confidant. If your particular build rule (`configure.js`) needs to read a config file, you can add it as an input and read it from within your rule. ie

``` js
module.exports = {
  // build task that depends on some configuration file
  {
    inputs: ['path/to/some/build-config.json'],
    ...
    rule: function() {
      var config = JSON.parse(require('fs').readFileSync('path/to/some/build-config.json', 'utf8'));
      // Now your task can use the contents of build-config.json
    }
  }
};
```

File: package.json [hunk 1]
```diff
@@ -5,26 +5,23 @@
   "bin": "./bin/confidant",
   "license": "MPL-2.0",
   "main": "./build/main",
-
```
Comment (lambdabaa) [inline package.json:8]: I like the newlines for readability

File: src/main.js [hunk 2]
```diff
@@ -1,7 +1,10 @@
+'use strict';
```
Comment (lambdabaa) [inline src/main.js]: babel adds this automatically

File: src/main.js [hunk 3]
```diff
@@ -1,7 +1,10 @@
...
+
 let ArgumentParser = require('argparse').ArgumentParser;
+let fs = require('fs');
```
Comment (lambdabaa) [inline src/main.js]: nit: Keep calls to `require` in alphabetical/ascii order

File: package.json [hunk 4]
```diff
@@ -5,26 +5,23 @@
...
   "dependencies": {
     "argparse": "^1.0.2",
     "babel": "^5.6.14",
     "debug": "^2.2.0",
     "glob": "^5.0.12",
     "lodash": "^3.9.3",
-    "mz": "^2.0.0"
+    "mz": "^2.0.0",
+    "nconf": "^0.7.1"
```
Comment (lambdabaa) [inline package.json]: I think we probably don't need `nconf`. We really just need to copy the stuff from `process.env` into our node script invocation

File: test/main_test.js [hunk 5]
```diff
@@ -72,7 +72,7 @@ suite('main', function() {
       .then(() => fs.unlink(`${dir}/build.ninja`))
       .catch(() => { /* will fail if state is already clean */ })
       .then(() => {
-        return main({ dir: dir });
+        return main({ dir: dir, config: __dirname + '/fixtures/config.json' });
```
Comment (lambdabaa) [inline test/main_test.js]: Can you add another test case in `test/fixtures` that makes sure that all of the environment variables are set correctly when the configure script gets run? One way to do that is to write a build rule that throws an error or exits with a nonzero status code if some env var's value is wrong.

File: package.json (hunk 1 above)
Comment (rickychien) [inline package.json:8]: I usually install package through command line:

```
$ npm install --save nconf  // for installing package
```

or

```
$ npm rm --save nconf  // for removing pakcage
```

so npm is so smart to know where is the right position to put `"nconf": "^0.7.1"`  in package.json automatically.
It also arrange package.json and remove redundant newlines.

Comment (lambdabaa) [inline package.json:8]: Okay fair enough

File: src/main.js [hunk 6]
```diff
@@ -1,7 +1,12 @@
+/**
```
Comment (lambdabaa) [inline src/main.js]: This is more of a project description and less useful as a file-level comment imo

File: src/main.js [hunk 7]
```diff
@@ -1,7 +1,12 @@
...
+ * @fileoverview Configure your project for ninja's parallel and incremental
+ *               builds with pure js.
+ */
+
 let ArgumentParser = require('argparse').ArgumentParser;
 let FindStream = require('./find_stream');
+let fs = require('fs');
```
Comment (lambdabaa) [inline src/main.js]: I think nothing in this file needs to change. Let's try to keep the changeset small.

File: src/ninja_stream.js [hunk 8]
```diff
@@ -15,13 +15,14 @@ function NinjaStream() {
   Transform.call(this, { objectMode: true });
   this.id = 0;
 
-  this.push(`# This ninja build file was generated by confidant.
+  this.push(`
```
Comment (lambdabaa) [inline src/ninja_stream.js]: This adds an unnecessary newline to the beginning of all of the build files

File: src/ninja_stream.js [hunk 9]
```diff
@@ -40,7 +41,19 @@ NinjaStream.prototype._transform = function(file, encoding, done) {
 NinjaStream.prototype._syncTransform = function(file, encoding, tasks, done) {
   let contents = readFileSync(file);
   let dir = dirname(file);
+
   debug(`Checking ${file}... found ${tasks.length} tasks`);
+
+  function envToString(env) {
```
Comment (lambdabaa) [inline src/ninja_stream.js]: nit: Can we define this outside of `NinjaStream.prototype._syncTransform`?

File: test/main_test.js [hunk 10]
```diff
@@ -13,14 +13,12 @@ let testCases = [
...
     verify: function() {
       let dir = this.dir;
-      return Promise.all(
-        [
-          `${dir}/build.ninja`,
-          `${dir}/random-copy`,
-          `${dir}/one/random-copy`,
-          `${dir}/other/random-copy`
-        ].map(checkExists)
-      );
```
Comment (lambdabaa) [inline test/main_test.js:23]: Same thing here. If you have stylistic things you want to address, let's move those bits into their own pull request to keep the history clean

File: test/main_test.js [hunk 11]
```diff
@@ -86,3 +91,14 @@ function checkExists(path) {
     `${path} should exist`
   );
 }
+
+function checkEnvInNinja(path) {
+  let read = new Promise((resolve, reject) => {
```
Comment (lambdabaa) [inline test/main_test.js]: I was thinking that you could even do this in your `env/configure.js`. Does that sound good?

Comment (rickychien) [inline test/main_test.js]: I prefer to change it by outputting a success file if a given env variable ENV_TEST=1 is existed with a correct value.

**Comment by rickychien** [comment]
@gaye 

In my experience, almost all build modules in current build system have to read config to determine how to do next step. (You can see that in gaia/build/*.js) So configure.js would be a pretty similar case and even we can't avoid requiring a config-manager in configure.js. If every configure.js has to `require('config-manager');`, why don't we just put it in confidant?
That's why I prefer to put config-manager inside confidant.

File: .gitignore [hunk 12]
```diff
@@ -10,6 +10,8 @@ node_modules
 /test/fixtures/config_filter/random-copy
 /test/fixtures/config_filter/one/random-copy
 /test/fixtures/config_filter/other/random-copy
+/test/fixtures/env/env
```
Comment (lambdabaa) [inline .gitignore]: This doesn't get created anymore?

File: src/main.js [hunk 13]
```diff
@@ -1,7 +1,7 @@
 let ArgumentParser = require('argparse').ArgumentParser;
+let createWriteStream = require('fs').createWriteStream;
```
Comment (lambdabaa) [inline src/main.js]: nit: Let's keep this in ascii order

**Comment by lambdabaa** [comment]
Other than a couple nits this LGTM. Nice!

**Comment by lambdabaa** [comment]
Landed on master. Thanks!
